src/lib/Simplify.hs

```diff
@@ -523,53 +525,61 @@ simplifyOp op = case op of
   Select c x y -> select c x y
   _ -> emitOp op
 
+pattern IdentityReconAbs :: Abs binder ReconstructAtom n
+pattern IdentityReconAbs <- Abs _ IdentityRecon
+
 simplifyHof :: Emits o => Hof i -> SimplifyM i o (Atom o)
 simplifyHof hof = case hof of
   For d lam@(Lam lamExpr) -> do
     ixTy <- substM $ argType lamExpr
-    (lam', recon) <- simplifyLam lam
+    (lam', Abs b recon) <- simplifyLam lam
     ans <- liftM Var $ emit $ Hof $ For d lam'
     case recon of
       IdentityRecon -> return ans
       LamRecon reconAbs ->
         buildTabLam "i" ixTy \i' -> do
           elt <- tabApp (sink ans) $ Var i'
-          applyReconAbs (sink reconAbs) elt
+          reconAbs' <- applySubst (b @> i') reconAbs
+          applyReconAbs reconAbs' elt
```

Comment:
It would be nice if we could avoid substituting the body of `recon` twice (once in `applySubst` and once in `applyReconAbs`). Perhaps we could make `applyReconAbs` run in a `SubstReader`?

Reply:
Left a todo note.  I don't know how to do this; want to pair it some time?